tests/openal_capture.c

```diff
@@ -0,0 +1,298 @@
+// This tests captures a fixed amount of audio data,
+// then plays it back.
+//
+// Wishlist:
+// - Try multiple devices simultaneously;
+// - Have several recording passes over the same fixed buffer_size.
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <assert.h>
+#include <unistd.h>
+#ifdef __EMSCRIPTEN__
+#include <emscripten.h>
+#define ASSUME_AL_FLOAT32
+#endif
+#include <AL/al.h>
+#include <AL/alc.h>
+#ifdef ASSUME_AL_FLOAT32
```

Comment:
Is it better to define theses constants in <AL/alext.h> ?
On my linux <AL/ext.h> I have this :
```
#ifndef AL_EXT_float32
#define AL_EXT_float32 1
#define AL_FORMAT_MONO_FLOAT32                   0x10010
#define AL_FORMAT_STEREO_FLOAT32                 0x10011
#endif
```

Reply:
Which constants do you mean ?
For `ASSUME_AL_FLOAT32`, it's not the same as `AL_EXT_float32` because the latter only indicates that `<AL/ext.h>` defines the relevant constants, which does not imply that the implementation actually supports the extension.  
For `AL_FORMAT_MONO_FLOAT32` and `AL_FORMAT_STEREO_FLOAT32`, I had decided that is was acceptable because these are only two constants which aren't going to change, ever, and allows us to skip a file inclusion (file which might not actually define these if it is too old (which I believe would be unlikely, but anyway)).

`ASSUME_AL_FLOAT32` represents the hard-coded knowledge that our implementation in Emscripten is guaranteed to support `AL_EXT_float32` at all times.

Reply:
I mean AL_FORMAT_MONO_FLOAT32 and AL_FORMAT_STEREO_FLOAT32.

Reply:
Ok. So yeah, see my rationale above. I do agree that it may "feel" kinda wrong, but since the effects are strictly the same, it even makes the replacement with `#include <AL/ext.h>` not worth doing.  

When more constants are needed, we'll probably want to do it, but right now it's fine as it is.

Reply:
Wait, I just realised that Emscripten doesn't have `<AL/ext.h>` (See [the AL folder](https://github.com/jpernst/emscripten/tree/incoming/system/include/AL)), so I think I understand better what you meant.

Am I missing anything ? (@jpernst @juj) We should create one if it's not been done already, and I wouldn't mind doing it.

Reply:
On my system it's `AL/alext.h`. I'm not sure if this is a standard header, since any extension related symbols can be obtained with `alGetEnumValue()` and such, but I'm not opposed to adding it.

Reply:
Yes its `AL/alext.h` in my system too.

Reply:
The `AL/alext.h` header file is not part of the OpenAL 1.1 Core SDK as downloaded on [the official (?) website](https://openal.org/downloads/) (Windows), and neither the spec nor the programming guide mention it (actually they don't mention any header file, unless I missed something).  
Said SDK does include, hovewer, `AL/efx.h`, `AL/efx-creative.h`, `AL/EFX-Util.h` and `AL/xram.h`, which contain preprocessor constants and function pointer typedefs not unlike `AL/alext.h`, which leads me to believe that the latter is a convenience provided by OpenAL-Soft.

Sure, one could say that OpenAL-Soft is the _de facto_ standard, but still not _standard_ in the sense we're interested in.

This further confuses me about what to do. On the one hand, providing an `AL/alext.h` (and friends ? Or not ?) would surely be convenient for Emscripten users, but on the other hand I feel it would be lying (and further propagating the false idea that `AL/alext.h` is standard).  
Furthermore:
1. As @jpernst mentioned, there are already standard ways to fetch extension-related items;
2. Emscripten's OpenAL implementation is likely to implement only a handful of extensions, now and in the future.

I would have added all those files and called it a day if I didn't think this was a big deal - so right now, I'm in favor of keeping things as-is, unless there's a strong reason not to.

Reply:
I don't have a particularly strong preference either way, but if `alext.h` isn't part of the main SDK, then I'm inclined to not bother.

Reply:
Yeah, looks like we don't have an `AL/alext.h` right now.

> We should create it if it's not been done already, and I wouldn't mind doing it.
> if alext.h isn't part of the main SDK, then I'm inclined to not bother.

> On the one hand, providing an AL/alext.h (and friends ? Or not ?) would surely be convenient for Emscripten users, but on the other hand I feel it would be lying (and further propagating the false idea that AL/alext.h is standard).

On these counts, we can go either way. In Emscripten we are taking a stance of compatibility over standards, and in general we are easy going about adding in semi-nonportable platform specific code, as long as it is not much of a burden that would conflict or confuse. (see library_x11.js and musl compat include directories as an example)

So even if a file `AL/alext.h` is not part of a documented standard, but e.g. OpenAL-Soft specific customization, then we can certainly add that file in, while making a comment in the file noting that it is OpenAL-Soft compatibility specific.

No need to do that in this PR, or if you don't need this functionality, then can leave it out altogether.

However, we would not want to add those files copied verbatim from OpenAL Soft or similar, but probably write them ground up specific to the supported extensions in Emscripten. (and especially if there are some licensing issues)